pkg/storage/registry/dynamic/dynamic.go

```diff
@@ -0,0 +1,174 @@
+}
+
+type config struct {
+	Rules      map[string]string `mapstructure:"rules"`
+	Rewrites   map[string]string `mapstructure:"rewrites"`
+	HomePath   string            `mapstructure:"home_path"`
+	DBUsername string            `mapstructure:"db_username"`
+	DBPassword string            `mapstructure:"db_password"`
+	DBHost     string            `mapstructure:"db_host"`
+	DBPort     int               `mapstructure:"db_port"`
+	DBName     string            `mapstructure:"db_name"`
+}
+
+// New returns an implementation of the storage.Registry interface that
+// redirects requests to corresponding storage drivers.
+func New(ctx context.Context, m map[string]interface{}) (storage.Registry, error) {
+	var c config
+	if err := cfg.Decode(m, &c); err != nil {
+		return nil, err
+	}
+
+	log := appctx.GetLogger(ctx)
+	annotatedLog := log.With().Str("storageprovider", "dynamic").Logger()
+
+	rt, err := initRoutingTree(c.DBUsername, c.DBPassword, c.DBHost, c.DBPort, c.DBName)
+	if err != nil {
+		return nil, errors.Wrap(err, "error initializing routing tree")
+	}
+
+	d := &dynamic{
+		c:   &c,
+		log: &annotatedLog,
+		r:   c.Rules,
+		rt:  rt,
+		ur: &rewriter.UserRewriter{
+			Tpls: c.Rewrites,
+		},
+	}
+
+	return d, nil
+}
+
+func initRoutingTree(dbUsername, dbPassword, dbHost string, dbPort int, dbName string) (*routingtree.RoutingTree, error) {
+	db, err := sql.Open("mysql", fmt.Sprintf("%s:%s@tcp(%s:%d)/%s", dbUsername, dbPassword, dbHost, dbPort, dbName))
+	if err != nil {
+		return nil, errors.Wrap(err, "error opening sql connection")
+	}
+
+	results, err := db.Query("SELECT path, mount_id FROM routing")
+	if err != nil {
+		return nil, errors.Wrap(err, "error getting routing table from db")
+	}
+
+	var rs []routingtree.Route
+
+	for results.Next() {
+		var r routingtree.Route
+		err = results.Scan(&r.Path, &r.MountID)
+		if err != nil {
+			return nil, errors.Wrap(err, "error scanning rows from db")
+		}
+		rs = append(rs, r)
+	}
+
+	return routingtree.New(rs), nil
+}
+
+// ListProviders lists all available storage providers.
+func (d *dynamic) ListProviders(ctx context.Context) ([]*registrypb.ProviderInfo, error) {
+	providers := []*registrypb.ProviderInfo{}
+	for p, a := range d.r {
+		providers = append(providers, &registrypb.ProviderInfo{
+			ProviderPath: p,
+			Address:      a,
+		})
+	}
+
+	return providers, nil
+}
+
+// GetHome returns the storage provider for the home path.
+func (d *dynamic) GetHome(ctx context.Context) (*registrypb.ProviderInfo, error) {
+	p, err := d.rt.GetProviders(d.c.HomePath)
+	if err != nil {
+		return nil, errors.New("failed to get home provider")
+	}
+
+	if a, ok := d.r[p[0]]; ok {
+		return &registrypb.ProviderInfo{
+			ProviderPath: d.c.HomePath,
+			Address:      a,
+		}, nil
+	}
+
+	return nil, errors.New("home not found")
+}
+
+// FindProviders returns the storage providers for a given ref.
+func (d *dynamic) FindProviders(ctx context.Context, ref *provider.Reference) ([]*registrypb.ProviderInfo, error) {
```

Comment:
It should now check if the `storageID` is there and return the provider address from the config mapping. Going with the path resolution will only happen otherwise.